core-tests/shared/src/test/scala/zio/ZIOSpec.scala

```diff
@@ -2115,6 +2161,40 @@ object ZIOSpec extends ZIOBaseSpec {
         }
       }
     ),
+    suite("validateM")(
+      testM("returns all errors if never valid") {
+        val in  = List.fill(10)(0)
+        val res = ZIO.validateM[Any, Int, Int, Int](in)(a => ZIO.fail(a)).flip
+        assertM(res, equalTo(in))
+      },
+      testM("accumulate errors and ignore successes") {
+        val in  = List.range(0, 10)
+        val res = ZIO.validateM(in)(a => if (a % 2 == 0) ZIO.effect(a) else ZIO.fail(a))
+        assertM(res.flip, equalTo(List(1, 3, 5, 7, 9)))
+      },
+      testM("accumulate successes") {
+        val in  = List.range(0, 10)
+        val res = IO.validateM(in)(a => ZIO.effect(a))
+        assertM(res, equalTo(in))
+      }
+    ),
+    suite("validateFirstM")(
+      testM("returns all errors if never valid") {
+        val in  = List.fill(10)(0)
+        val res = ZIO.validateFirstM[Any, Int, Int, Int](in)(a => ZIO.fail(a)).flip
+        assertM(res, equalTo(in))
+      },
+      testM("short circuits on first success validation") {
+        val in = List.range(1, 10)
+        val f  = (a: Int) => if (a == 5) ZIO.effect(a) else ZIO.fail(a)
```

Comment:
Let's use a number other than the middle so the test will fail if we don't evaluate the effects in the correct order.